packages/model-viewer/src/features/annotation.ts

```diff
@@ -0,0 +1,208 @@
+    const positionNodes = parseExpressions(element.dataset.position!)[0].terms;
+    for (let i = 0; i < 3; ++i) {
+      this.position.setComponent(
+          i, normalizeUnit(positionNodes[i] as NumberNode<'m'>).number);
+    }
+  }
+
+  updateNormal(element: HTMLElement) {
+    const normalNodes = parseExpressions(element.dataset.normal!)[0].terms;
+    for (let i = 0; i < 3; ++i) {
+      this.normal.setComponent(
+          i, normalizeUnit(normalNodes[i] as NumberNode<'m'>).number);
+    }
+  }
+}
+
+export declare interface AnnotationInterface {
+  hiddenOpacity: number;
+  hiddenAngle: string;
+  addHotspot(element: HTMLElement): void;
+}
+
+export const AnnotationMixin = <T extends Constructor<ModelViewerElementBase>>(
+    ModelViewerElement: T): Constructor<AnnotationInterface>&T => {
+  class AnnotationModelViewerElement extends ModelViewerElement {
+    @property({type: Number, attribute: 'hidden-opacity'})
+    hiddenOpacity: number = DEFAULT_HIDDEN_OPACITY;
+
+    @style({
+      intrinsics: {
+        basis: [numberNode(DEFAULT_HIDDEN_ANGLE, 'rad')],
+        keywords: {auto: [null]}
+      },
+      updateHandler: $syncHiddenAngle
+    })
+    @property({type: String, attribute: 'hidden-angle'})
+    hiddenAngle: string = 'auto';
+
+    private[$annotationRenderer] = new CSS2DRenderer();
+    private[$hiddenAngle] = DEFAULT_HIDDEN_ANGLE;
+    private[$nextHotspotIndex] = 0;
+    private[$hotspotMap] = new Map();
+    private[$observer] = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        switch (mutation.type) {
+          case 'childList':
+            mutation.addedNodes.forEach((node) => {
+              this.addHotspot(node);
+            });
+            mutation.removedNodes.forEach((node) => {
+              this.removeHotspot(node);
+            });
+            break;
+          case 'attributes':
+            const {target} = mutation;
+            const hotspot = this[$hotspotMap].get(target);
+            if (!hotspot) {
+              this.addHotspot(target);
+              break;
+            }
+            switch (mutation.attributeName) {
+              case 'data-position':
+                hotspot.updatePosition(target);
+                break;
+              case 'data-normal':
+                hotspot.updateNormal(target);
+                break;
+            }
+            break;
+        }
+      });
+    });
+
+    connectedCallback() {
+      super.connectedCallback();
+      const {domElement} = this[$annotationRenderer];
+      const {style} = domElement;
+      style.pointerEvents = 'none';
+      style.position = 'absolute';
+      style.top = '0';
+      this.shadowRoot!.querySelector('.container')!.appendChild(domElement);
+
+      for (let i = 0; i < this.children.length; ++i) {
+        this.addHotspot(this.children[i]);
+      }
+
+      this[$observer].observe(
+          this, {childList: true, attributes: true, subtree: true});
```

Comment:
`subtree: true` is both slow and probably not what you want; it makes the observer recursive, so it will notify you of all such changes (child list and attribute changes) to both the observed node and each of its descendants.

Reply:
Yeah, the reason I put this in is because I wanted to be notified when the immediate children's attributes changed, but all I was getting was the containers' attributes.

Reply:
Yah, for the first pass let's skip observing changes to the attributes. There isn't really a good way to do it without either A) requiring hotspots to be custom elements or B) observing each node individually with a second mutation observer.